### PR TITLE
Fapi ciba/encode decode login hint

### PR DESCRIPTION
**Pull request by DmitryMishchuk**

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java [hunk 1]
```diff
@@ -1,19 +1,35 @@
...
+    private static final Logger logger = Logger.getLogger(DelegateDecoupledAuthenticationProvider.class);
 
     public DefaultCIBALoginUserResolver(KeycloakSession session) {
         this.session = session;
     }
 
     @Override
     public UserModel getUserFromLoginHint(String loginHint) {
+        String secret = session.getContext().getClient().getSecret();
+        if (StringUtils.isNotBlank(secret)) {
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java:25]: It causes the failures of all tests of existing arquillian integration test `CIBATest`.

IMO, it might to be better to add the Client setting `login_hint encryption` switch.

### Where to add new setting item on Admin Console
IMO, on current client's "Settings" UI, add this `login_hint encryption` switch on existing "Advanced Settings" pain.

### Where to look if we  introduce new client setting 

Sometimes we need to introduce new Client setting.

There are two types of Client setting.

- OAuth/OIDC standard document specified Client Metadata
  - OAuth2
     https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml
  - OIDC
     https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata
- NOT OAuth/OIDC standard document specified Client Metadata

AFAIK, entities we need to look at when we introduce new client setting (regardless of whether it is OAuth/OIDC standard document specified Client Metadata or not) are as follows.

- Representation
 It stands for the format in representing Client.
 This format is used when communicating between keycloak and other entities.
   - [1] ClientRepresentation
 It is Client's Representation.
 When we access Client by Admin REST API(namely, on Admin Console), this ClientRepresentation is used.

- Model
 It stands for the format used when keycloak's internal program treats Client.
  - [2]  ClientModel
 It is Client's Model

- Store
 It stands for the format used when keycloak's internal program stores Client.
   - Infinispan Cache
  Used when storing Client in memory
      - [3] ClientAdapter(infinispan)
   - JPA:
  Used when storing Client in persistent storage
      - [4] ClientAdapter(jpa)
      - [5] ClientEntity(jpa)

- OIDC Extension
 It extends Client so that it can be treated as OIDC conformant Client
   - Representation
     - [6] OIDCClientRepresentation
   It stands for the format in representating OIDC conformant Client.
   This format is used when communicating between keycloak and other entities in OIDC manner
   eg.
   When we access Client by OIDC Dynamic Client Registration, this OIDCClientRepresentation is used.

   - Representation/Model
      - [7] OIDCAdvancedConfigWrapper
    It works as the wrapper between ClientRepresentation and OIDCClientRepresentation
      - [8] DescriptionConverter
    It converts the Client representation between ClientRepresentation and OIDCClientRepresentation.
      - [9] OIDCConfigAttributes
     It define keys used to store/access OIDC related parameter to ClientRepresentation's map object called Attribute and ClientModel's map object also callled Attribute.

- UI
   - [10] client.js
   - [11] client-detail.html
   - [12] admin-messages_en.properties 

Basically,  we need not modify [3][4][5].
We need to modify [6][7] if we introduce new OAuth/OIDC standard document specified Client Metadata.

#### For OAuth/OIDC standard document specified Client Metadata
We need to modify [1][2][6][7][8][9][10][11][12]. 

#### For NOT OAuth/OIDC standard document specified Client Metadata
We need to modify [1][2][8][9][10][11][12].

### Reference
The PRs about newly introducing Client settings are as follows :

- https://github.com/keycloak/keycloak/pull/6581
- https://github.com/keycloak/keycloak/pull/7065

File: services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        return getCipher(Cipher.DECRYPT_MODE, secret).doFinal(encryptedBytes);
+    }
+
+    private static Cipher getCipher(int mode, String secret) throws GeneralSecurityException {
+        UUID uuid = UUID.fromString(secret);
+        byte[] bytes = ByteBuffer.wrap(new byte[AUTH_TAG_SIZE_BYTE])
+                               .putLong(uuid.getMostSignificantBits())
+                               .putLong(uuid.getLeastSignificantBits())
+                               .array();
+        SecretKeySpec secretKey = new SecretKeySpec(bytes, "AES");
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:33]: AFAIK, keycloak tends to use BC (BouncyCastle) as security provider, not SunJCE.

https://github.com/tnorimat/keycloak/blob/be4780243b702e9b7e8aa236bbde1312b93ebfea/core/src/main/java/org/keycloak/jose/jwe/enc/AesGcmEncryptionProvider.java#L91

https://github.com/tnorimat/keycloak/blob/be4780243b702e9b7e8aa236bbde1312b93ebfea/core/src/main/java/org/keycloak/jose/jwe/enc/AesGcmEncryptionProvider.java#L101

File: services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        return getCipher(Cipher.ENCRYPT_MODE, secret).doFinal(bytes);
+    }
+
+    public static byte[] decrypt(String secret, byte[] encryptedBytes) throws GeneralSecurityException {
+        return getCipher(Cipher.DECRYPT_MODE, secret).doFinal(encryptedBytes);
+    }
+
+    private static Cipher getCipher(int mode, String secret) throws GeneralSecurityException {
+        UUID uuid = UUID.fromString(secret);
+        byte[] bytes = ByteBuffer.wrap(new byte[AUTH_TAG_SIZE_BYTE])
```
Comment (tnorimat) [inline services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:29]: I'm not sure whether AAD is required or not in this case.
Is it safe to encrypt/decrypt data in AES-GCM without AAD in this case?

**Review by tnorimat — CHANGES_REQUESTED** [review]
@DmitryMishchuk Thank you for your PR. I've reviewed it and commented on it. Could you please check them?

File: services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java (hunk 3 above)
Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:29]: In general it should not influence security of the encryption/decryption process greatly.
As I've seen in documentation if AAD is used it should be re-generated for each pair encrypt/decrypt operation, but then we will need to store it somewhere, like in existing code ADD of JWE it is stored in base64Header of JWE, here I did not find a place to store or get the AAD from in our flow, that is why I simply mocked it, we could add corresponding bytes to the. beginning or the end of the login_hint, but that would mean anybody can get it from there and there actually is no reason to have it as such. Any suggestion from your side?

Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:29]: And AAD is required for decryption by the Cipher.init(), so can't simply remove it.

File: services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java (hunk 1 above)
Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java:25]: So maybe here it would be better just check it for null and not for being blank assuming all our clients should be private and have to have a client_secret?

Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java:25]: By adding a setting I suppose you mean add new flag to client configuration to check it afterwards from within our code for like if/else statement?

Comment (tnorimat) [reply services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java:25]: Yes, e.g. "CIBA Login Hint Encryption" switch whose boolean value is stored on Client's attribute.

File: services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java (hunk 3 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:29]: I see, in this case, IV and AAD can not be shared among the client and keycloak by `login_hint` itself unlike JWE whose format includes IV and AAD. Thank you.

Comment (tnorimat) [reply services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:29]: I see, in this case, IV and AAD can not be shared among the client and keycloak by `login_hint` itself unlike JWE whose format includes IV and AAD. Thank you.

File: services/src/test/java/org/keycloak/utils/AesGcmEncryptorTest.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package org.keycloak.utils;
```
[the remark this answers is not in the document]
Comment (DmitryMishchuk) [reply services/src/test/java/org/keycloak/utils/AesGcmEncryptorTest.java:1]: Yes I will, this was an easy and fast way to test it is working as was expected.

File: services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java (hunk 2 above)
Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java:35]: updated to use BouncyCastle

File: services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java (hunk 1 above)
Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java]: Added switch on FE and corresponding checks in BE ([6-12]). Still didn't get why do we need to change [1] & [2], so advice would be highly appreciated.

Comment (DmitryMishchuk) [reply services/src/main/java/org/keycloak/protocol/ciba/resolvers/DefaultCIBALoginUserResolver.java]: StringUtils removed from code.

File: services/src/test/java/org/keycloak/utils/AesGcmEncryptorTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (DmitryMishchuk) [reply services/src/test/java/org/keycloak/utils/AesGcmEncryptorTest.java]: Tests added.

[the remark this answers is not in the document]
Comment (tnorimat) [reply services/src/test/java/org/keycloak/utils/AesGcmEncryptorTest.java]: I've confirmed. Thank you.

**Review by tnorimat — APPROVED** [review]

**Comment by tnorimat** [comment]
@DmitryMishchuk Thank you very much!

**Comment by tnorimat** [comment]
@DmitryMishchuk 
I'd like to add `Co-authored-by: ` trailer when upstreaming this CIBA support to keycloak to show your contribution explicitly.
Could you please confirm that  your author information is the following?
`Dmytro Mishchuk <dmi@adorsys.com.ua>`

**Comment by DmitryMishchuk** [comment]
> @DmitryMishchuk
> I'd like to add `Co-authored-by: ` trailer when upstreaming this CIBA support to keycloak to show your contribution explicitly.
> Could you please confirm that your author information is the following?
> `Dmytro Mishchuk <dmi@adorsys.com.ua>`

Yes that is correct,
Regards
